pkg/operator/controller/dns_status.go

```diff
@@ -0,0 +1,173 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncDNSStatus computes the current status of dns and
+// updates status upon any changes since last sync.
+func (r *reconciler) syncDNSStatus(dns *operatorv1.DNS, clusterIP, clusterDomain string, ds *appsv1.DaemonSet) error {
+	updated := dns.DeepCopy()
+	updated.Status.ClusterIP = clusterIP
+	updated.Status.ClusterDomain = clusterDomain
+	updated.Status.Conditions = computeDNSStatusConditions(dns.Status.Conditions, clusterIP, ds)
+	if !dnsStatusesEqual(updated.Status, dns.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update dns status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// computeDNSStatusConditions computes dns status conditions
+// based on the status of ds and clusterIP.
+func computeDNSStatusConditions(conditions []operatorv1.OperatorCondition, clusterIP string, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
+	conditions = setDNSStatusCondition(conditions, computeDNSDegradedCondition(clusterIP, ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSProgressingCondition(ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSAvailableCondition(clusterIP, ds))
+	return conditions
+}
+
+// computeDNSDegradedCondition computes the dns Degraded status condition
+// based on the status of clusterIP and ds.
+func computeDNSDegradedCondition(clusterIP string, ds *appsv1.DaemonSet) *operatorv1.OperatorCondition {
+	degradedCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.OperatorStatusTypeDegraded,
+		Status: operatorv1.ConditionUnknown,
+	}
+	switch {
+	case len(clusterIP) == 0 && ds.Status.NumberAvailable == 0:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "NoClusterIPAndDaemonSet"
+		degradedCondition.Message = "No ClusterIP for DNS Service and no DaemonSet pods running"
+	case len(clusterIP) == 0:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "NoClusterIP"
+		degradedCondition.Message = "No Cluster IP for DNS Service available"
+	case ds.Status.NumberAvailable == 0:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "DaemonSetUnavailable"
+		degradedCondition.Message = "DaemonSet pod not running on any nodes"
+	case ds.Status.NumberAvailable != ds.Status.DesiredNumberScheduled:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "DaemonSetDegraded"
+		degradedCondition.Message = "All nodes not running DaemonSet pod"
```

Comment:
This could be misinterpreted as saying that no nodes are running the DaemonSet pod.  "Not all nodes running" or "Some nodes not running" would be clearer.